### ISSUE TITLE
F-Droid inclusion

### ISSUE THREAD

**Issue by Wv5twkFEKh54vo4tta9yu7dHa3** [admin-ch/CovidCertificate-App-Android#40]
Hi, is it OK if we add this app to F-Droid? It seems to be fully FOSS so it would be good to be on there. 

Would you accept a PR to add Fastlane metadata? Cf https://gitlab.com/snippets/1895688

Thanks for your work and for releasing it as free software :)

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
F-Droid inclusion merge request here: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/9099

This just adds the wallet app for now, will also add the verifier if that goes well :)

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
We are trying to get reproducible builds to work (so that APKs published on F-Droid are published with your signatures). It would be easier if you could publish the buildTimestamp in the release notes, along with the APK. Would that be possible for you?

**Comment by goebelUB** [comment]
Thanks a lot for your interest! We appreciate your effort and motivation to help.

We are currently actively discussing alternative publishing methods (F-Droid, ...) with the project team. The publishing process and go-live need to be aligned between all platforms. We will keep you posted once an official decision is reached.

For the time being, the app is and should not yet be available on any stores. This is also due to ongoing tests (including the [public security test](https://www.ncsc.admin.ch/ncsc/en/home/dokumentation/covid-certificate-pst/infos.html)).

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Thanks for the reply, of course the inclusion in F-Droid is on hold until the app gets published, I just wanted to get a head start to not have too much of a delay in publication there.

Thanks for not dismissing this out of hand, the fact that you're even considering it is much better than responses for other apps :) :heart:

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Hey @goebelUB , since the app is now published on Google Play, can we move forward with publication on the F-Droid side?

We would also like to try to get reproducible builds to work. In the future, would it be possible to publish the build timestamps in the release notes?

And again, would you accept a PR to add fastlane metadata to the repo? Or even better, could you add your own fastlane metadata (since those will have translations of the app store descriptions etc)

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Thanks for adding the build timestamps to the releases :heart:

**Comment by goebelUB** [comment]
You are very quick to pick this up! I'll try to make sure that build timestamps are added to any future releases, too.

I will also add the metadata, though [it seems](https://f-droid.org/en/docs/All_About_Descriptions_Graphics_and_Screenshots/#in-the-applications-source-repository) that it'll have to be in Triple-T structure in order to support two apps in the same repo.

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
>  it seems that it'll have to be in Triple-T structure in order to support two apps in the same repo

Yes indeed.

> You are very quick to pick this up

Well, I was on the releases page, because I was looking to get reproducible builds to work.
They are not working out of the gate, it seems subtle differences in build environments exist. Let me attach a diffoscope result, for future reference.

[diffoscope.txt](https://github.com/admin-ch/CovidCertificate-App-Android/files/6622514/diffoscope.txt)

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Also thanks for the PR to fix the gradle checksum missing (#76)

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
From the diffoscope it seems the source of the tag doesn't correspond to the apk published... I think the UserAgent stuff is different, maybe? It's hard to see if this concerns generated code or actual code you wrote. Some stuff fails to decompile with the diffoscope. Maybe it's just because of build systems differences, idk

I doubt this is going to be easy :sweat_smile:

**Comment by goebelUB** [comment]
I see, the branching around this specific wallet release might have messed things up, I'll look into it.

In the meantime, can you see if F-Droid can reproduce [this verifier release](https://github.com/admin-ch/CovidCertificate-App-Android/releases/tag/v1.0.1-versionCode-1010-verifier)? Maybe that works...

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Sure, I will make a draft MR for the verifier as well so we can test that.

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Something I noticed while adding the verifier to F-Droid: the apk attached to the release has the versioncode at the end (it's of the form wallet-prod-%v-%c-signed.apk), whereas the wallet apk attached to the release doesn't (it's of the form verifier-prod-%v-signed.apk).

Is this going to continue to be the case? If the format changes once it's included, it will break how F-Droid fetches the apk to extract the signature

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Holy shit it worked! https://gitlab.com/artectrex/fdroiddata/-/jobs/1337130549

:partying_face:

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
I will be moving forward with publishing the APK for the verifier on F-Droid. Will publish the wallet as soon as you publish the next release - where the apk will hopefully correspond to the matching tag this time :stuck_out_tongue_closed_eyes:

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Indeed, that's how it goes, we'll have to wait for the next release.

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
Both of them are now ready for review and merging into fdroid-data, I am waiting for the full release to mark the PRs as ready to be merged.

Both have reproducible builds, working great, and it's possible that the automatic updates will work (which we will see at the next update, otherwise I will have to do manual updates, but it should be fine?)

**Comment by obfusk** [comment]
Adding a ["get it on F-Droid" badge](https://gitlab.com/fdroid/artwork/tree/master/badge) next to the "get it on Google Play" one -- once the apps are available in F-Droid -- would be nice :)

**Comment by Wuestenschiff** [comment]
@Wv5twkFEKh54vo4tta9yu7dHa3 @goebelUB Looks like metadata is in 1.2 Release so can we get it now on F-droid?

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
It's ready to merge, no clue what is holding things up on the f-droid side... Someone with the correct rights needs to hit a button to merge it

**Comment by supaeasy** [comment]
Why is this not approved yet?

**Comment by goebelUB** [comment]
https://gitlab.com/fdroid/fdroiddata/-/merge_requests/9099#note_607719783

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
They don't have time, apparently? Really don't know what is required, imo it just needs someone to push the merge button... 

https://gitlab.com/fdroid/fdroiddata/-/merge_requests/9099#note_607719783

**Comment by Wuestenschiff** [comment]
> Why is this not approved yet?

F-Droid is just super slow / overlaoded. You need to find someone with F-Droid merge right and nicly ask him to merge ist. Everything is done on this repo

**Comment by supaeasy** [comment]
It has been merged. Still no Idea why it does not show up in F-Droid for me.

**Comment by jspricke** [comment]
it needs to be build and published still, you can see the current state here: https://monitor.f-droid.org/builds/running
Usually a build cycle takes approx. three days, so normally it should be published in 4 to 5 days but currently it is a little stuck.

**Comment by supaeasy** [comment]
> it needs to be build and published still, you can see the current state here: https://monitor.f-droid.org/builds/running
> Usually a build cycle takes approx. three days, so normally it should be published in 4 to 5 days but currently it is a little stuck.

Thanks a lot. I was not familiar with this.

**Comment by goebelUB** [comment]
v.1.2.0 is now on F-Droid (visible in the client after an index update, but not yet visible on the website).
It seems that the auto update worked as well ([verifier](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/ch.admin.bag.covidcertificate.verifier.yml), [wallet](https://gitlab.com/fdroid/fdroiddata/-/blob/master/metadata/ch.admin.bag.covidcertificate.wallet.yml)).

Unfortunately, the metadata is flipped: what looks like the wallet downloads the verifier, and vice versa:
<img src=https://user-images.githubusercontent.com/56064810/124284608-6f3a7980-db4d-11eb-8fac-3c3301d6db7b.jpg width=30%>
I don't yet understand why, the metadata in the repo looks fine, e.g. see the app titles: [wallet](https://github.com/admin-ch/CovidCertificate-App-Android/blob/v1.2.0-1200-wallet/wallet/src/main/play/listings/en-US/title.txt), [verifier](https://github.com/admin-ch/CovidCertificate-App-Android/blob/v1.2.0-1200-verifier/verifier/src/main/play/listings/en-US/title.txt).

**Comment by obfusk** [comment]
> Unfortunately, the metadata is flipped: what looks like the wallet downloads the verifier, and vice versa:

I think this is a bug in f-droid. Will investigate.

**Comment by obfusk** [comment]
> I think this is a bug in f-droid. Will investigate.

Looks like it. [Working on a fix.](https://gitlab.com/fdroid/fdroidserver/-/merge_requests/981)

**Comment by obfusk** [comment]
FYI: if you add a changelog to the metadata, the app will show up in "latest apps" on F-Droid.

**Comment by goebelUB** [comment]
>  Working on a fix.

Thanks a lot for the quick response!

>  if you add a changelog to the metadata, the app will show up in "latest apps" on F-Droid.

I know that this is the last missing straw, all other boxes are ticked (translations, graphics, ...). v2.0.0 contains a changelog.

**Comment by obfusk** [comment]
> Thanks a lot for the quick response!

Getting the fix into production might take a little while though (certainly not until after the weekend).

**Comment by obfusk** [comment]
lol. v2.0.0 is now in the repo. and the nondeterminism just happened to produce the correct result this time.

**Comment by goebelUB** [comment]
Well, mostly. The changelogs are still switched. And interestingly the screenshots are correct on the website, but (at least for me) flipped in the client. Also on both, the verifier - which should only have 3 screenshots - shows the last 2 screenshots from the wallet too.

**Comment by obfusk** [comment]
> That's a bug I've seen before :) Will report.

So that's a [known bug](https://gitlab.com/fdroid/fdroidserver/-/issues/490). I'll see if I can so something about it soon, but it's not trivial to fix.

**Comment by goebelUB** [comment]
A heads up: in the new version we [pulled out the core SDK](https://github.com/admin-ch/CovidCertificate-SDK-Kotlin) to share it with the backend.
We'd like to have it on Maven, unfortunately this has been slow. In the meantime we've [included it as a JAR](https://github.com/admin-ch/CovidCertificate-SDK-Android/tree/main/sdk/libs).
I believe the F-Droid scanner may not be happy about this. What would be the best options here?

Btw I'll keep this issue open to keep all F-Droid discussions in one place. Anybody no longer interested in this issue can always unsubscribe from Github's notifications. At the same time anybody interested will get notifications without being @ ed.

**Comment by Wv5twkFEKh54vo4tta9yu7dHa3** [comment]
That's not OK for F-Droid, it will need to stop builds. We could maybe build this from source, but I don't know if it will be reproducible. Won't have time for this for the coming month though... Someone else to look at this?

**Comment by obfusk** [comment]
> A heads up: in the new version we [pulled out the core SDK](https://github.com/admin-ch/CovidCertificate-SDK-Kotlin) to share it with the backend.
> We'd like to have it on Maven, unfortunately this has been slow. In the meantime we've [included it as a JAR](https://github.com/admin-ch/CovidCertificate-SDK-Android/tree/main/sdk/libs).
> I believe the F-Droid scanner may not be happy about this. What would be the best options here?

Maven would be fine afaik. A jar (or any other binary) is not acceptable for f-droid.

You could use a separate build flavour for f-droid that builds the .jar from source.

Feel free to ping me if I can help.

**Comment by goebelUB** [comment]
In that case I think the easiest solution is for F-Droid to skip/disable v2.1.0. The next app release should be without the JAR again and [via Maven](https://github.com/admin-ch/CovidCertificate-SDK-Kotlin/pull/26) instead.

**Comment by Wuestenschiff** [comment]
@goebelUB Also the button for the ligth certificate is missing on my phone, is this also a bug of the f-droid build? According to change log it should be in in version 2.0

**Comment by goebelUB** [comment]
F-Droid builds are reproducible, i.e. the apk you get from F-Droid is exactly the same as the apk distributed via Google Play, AppGallery and Github. This also allows you yo upgrade across stores. So there should not be any difference.

2.0.0 added transfer codes, the light certificate was only added in 2.1.0.

**Comment by Wuestenschiff** [comment]
> 2.0.0 added transfer codes, the light certificate was only added in 2.1.0.
Then there is a bug in the metadata, f-droid says:
```
NEW IN VERSION 2.0.0

- Certificate light: QR ....
```

**Comment by goebelUB** [comment]
Here's the metadata for the wallet [2.0.0](https://github.com/admin-ch/CovidCertificate-App-Android/blob/v2.0.0-2000-wallet/wallet/src/main/play/release-notes/en-US/default.txt) and [2.1.0](https://github.com/admin-ch/CovidCertificate-App-Android/blob/v2.1.0-2100-wallet/wallet/src/main/play/release-notes/en-US/default.txt).
Note that in 2.0.0 I accidentally swapped the changelogs, so see also the verifier [2.0.0](https://github.com/admin-ch/CovidCertificate-App-Android/blob/v2.0.0-2000-wallet/verifier/src/main/play/release-notes/en-US/default.txt).
The [releases page](https://github.com/admin-ch/CovidCertificate-App-Android/releases) is also linked from F-Droid I believe, just in case.

This looks like a bug in F-Droid server, where it takes the changelog from newer metadata. It tried to build 2.1.0, [failed](#206), but still updated the metadata.

**Comment by obfusk** [comment]
> This looks like a bug in F-Droid server, where it takes the changelog from newer metadata. It tried to build 2.1.0, failed, but still updated the metadata.

See https://gitlab.com/fdroid/fdroidserver/-/issues/919

**Comment by obfusk** [comment]
> Unfortunately, the metadata is flipped: what looks like the wallet downloads the verifier, and vice versa:

This *should* be fixed from now on (currently both apps have the wallet metadata). Please let me know if it's not fixed after the next release is published.

**Comment by dR3b** [comment]
Status?

**Comment by goebelUB** [comment]
2.7.0 is now available on F-Droid. Sorry for the delay!

We ended up needing a separate release for F-Droid. That's not ideal, so I'll track this in #304.
